arch/lkl/kernel/irq.c

```diff
@@ -45,24 +46,54 @@ static inline void set_irq_status(int irq)
 	__sync_fetch_and_or(&irq_index_status, BIT(index));
 }
 
-
 static struct irq_info {
 	const char *user;
 } irqs[NR_IRQS];
 
+static bool irqs_enabled;
+
+static void run_irq(int irq)
+{
+	unsigned long flags;
+
+	/* interrupt handlers need to run with interrupts disabled */
+	local_irq_save(flags);
+	irq_enter();
+	generic_handle_irq(irq);
+	irq_exit();
+	local_irq_restore(flags);
+}
 
 /**
- * DO NOT run any linux calls (e.g. printk) here as they may race with the
- * existing linux threads.
+ * This function can be called from arbitrary host threads, so do not
+ * issue any Linux calls (e.g. prink) if lkl_cpu_get() was not issued
+ * before.
  */
 int lkl_trigger_irq(int irq)
 {
+	int ret;
+
 	if (!irq || irq > NR_IRQS)
 		return -EINVAL;
 
-	set_irq_status(irq);
+	ret = lkl_cpu_try_run_irq(irq);
+	if (ret <= 0)
+		return ret;
+
+	/*
+	 * Since this can be called from Linux context (e.g. lkl_trigger_irq ->
+	 * IRQ -> softirq -> lkl_trigger_irq) make sure we are actually allowed
+	 * to run irqs at this point
+	 */
+	if (!irqs_enabled) {
+		set_irq_pending(irq);
```

Comment:
In such a case, when does the irq run? Should we also set cpu.irq_pending so that the outmost lkl_cpu_put can run the irq?